core/src/main/java/io/paradoxical/cassieq/clustering/election/HazelcastLeadershipProvider.java

```diff
@@ -0,0 +1,152 @@
+
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+
+import static com.godaddy.logging.LoggerFactory.getLogger;
+import static java.util.stream.Collectors.toSet;
+
+public class HazelcastLeadershipProvider implements LeadershipProvider {
+    private final HazelcastInstance hazelcastInstance;
+    private final ClusteringConfig config;
+
+    private static final Logger logger = getLogger(HazelcastLeadershipProvider.class);
+
+
+    @Inject
+    public HazelcastLeadershipProvider(
+            HazelcastInstance hazelcastInstance,
+            ClusteringConfig config) {
+        this.hazelcastInstance = hazelcastInstance;
+        this.config = config;
+    }
+
+    @Override
+    public boolean tryAcquireLeader(final LeadershipRole key) {
+        return tryAquireLock(key, keySetter -> {
+            final ClusterMember currentOwner = keySetter.value();
+
+            // its already us
+            if (currentOwner != null && currentOwner.equals(keySetter.localId())) {
+                return true;
+            }
+
+            // stale or unclaimed
+            if (currentOwner == null || !inCluster(currentOwner)) {
+                keySetter.claim();
+
+                return true;
+            }
+
+            // its someone else
+            return false;
+        });
+    }
+
+    @Override
+    public boolean tryRelinquishLeadership(final LeadershipRole key) {
+        return tryAquireLock(key, keySetter -> {
+            if (keySetter.amLeader()) {
+                keySetter.release();
+
+                return true;
+            }
+
+            return false;
+        });
+    }
+
+    private boolean inCluster(final ClusterMember clusterId) {
+        return hazelcastInstance.getCluster()
+                                .getMembers()
+                                .stream()
+                                .map(Member::getUuid)
+                                .map(ClusterMember::valueOf)
+                                .collect(toSet())
+                                .contains(clusterId);
+    }
+
+    private boolean tryAquireLock(final LeadershipRole key, final Function<KeySetter, Boolean> action) {
+        final KeySetter keySetter = new KeySetter(key);
+
+        if (keySetter.tryLock()) {
+            try {
+                return action.apply(keySetter);
+            }
+            finally {
+                keySetter.unlock();
+            }
+        }
+
+        return false;
+    }
+
+    class KeySetter {
+        private final IMap<LeadershipRole, String> backingMap;
+        private final LeadershipRole key;
+
+        public KeySetter(LeadershipRole key) {
+            this.backingMap = hazelcastInstance.getMap("leaders");
+
+            this.key = key;
+        }
+
+        private ClusterMember localId() {
+            return ClusterMember.valueOf(hazelcastInstance.getCluster().getLocalMember().getUuid());
+        }
+
+        public boolean amLeader() {
```

Comment:
this might have to take in a clusterMember and be called `isLeader(ClusterMember)` if the above method is moved